### PR TITLE
Rework pwise scheduler for better broadcasting perf on TIMM VIT.

### PR DESCRIPTION
**Pull request by csarofeen**
xSpeedup of change is in far right column. No significant regression seen with this change.
image

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 1]
```diff
@@ -326,14 +343,15 @@ c10::optional<PointwiseParams> getPointwiseHeuristics(
...
   // point.
-  int64_t gdimx = 1;
+  int64_t gdim_left = 1;
 
-  // gdimy is used if there's too much parallelization in the right side of the
-  // break point. We will expand grid parallelization into the right side of the
-  // break point with gdimx and use gdimy for the left side of the break point.
-  int64_t gdimy = 1;
+  // gdim_right is used if there's too much parallelization in the right side of
+  // the break point. We will expand grid parallelization into the right side of
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:351]: > We will expand grid parallelization into the `left` side of the break point with gdim_left and use gdim_right for the `right` side of the break point.

Should the comment be changed to this ^^^?

**Comment by csarofeen** [comment]
@rdspring1 @naoyam I'm using the admin hammer to get this PR in the nightly tonight. Please retroactively review and I will follow up with any requested changes.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 2]
```diff
@@ -392,51 +408,63 @@ c10::optional<PointwiseParams> getPointwiseHeuristics(
...
-        // Don't limit unroll factor with break point
-        if (cur_right_elem_count < max_unroll_factor) {
+        // Need to be able to parallelize, don't use break if there's not
+        // at least an unrolled warp.
+        if (ceilDiv(cur_right_elem_count, max_unroll_factor) <=
+            at::cuda::getCurrentDeviceProperties()->warpSize) {
           continue;
         }
 
+        // If outer broadcast, or balanced broadcast:
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:438]: Refactor suggestion:
```c++
// If outer broadcast or balanced broadcast AND right transfer size is bigger than half of L2
// then flip BIDx and BIDy bindings
flip_grid_binding = (lhs_byte_multiple <= rhs_byte_multiple) &&
                    (at::cuda::getCurrentDeviceProperties()->l2CacheSize < right_transfer_size * 2);
```
The If-Else conditional seems superfluous.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 3]
```diff
@@ -736,26 +760,60 @@ void schedulePointwise(Fusion* fusion, const PointwiseParams& params) {
     //[i-remainder | outer | Unswitch, Unroll, TIDx]
     if (params.split_block) {
       reference_tv->split(1, NamedScalar::getParallelDim(ParallelType::TIDy));
-      // [i-remainder | BIDx TIDy | Unswitch, Unroll, TIDx]
-      reference_tv->axis(1)->parallelize(ParallelType::BIDx);
-      reference_tv->axis(2)->parallelize(ParallelType::TIDy);
+      if (params.flip_grid_binding) {
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:763]: Refactor suggestion:
```c++
        // Before: [i-remainder | outer | Unswitch, Unroll, TIDx]
        // If split_block:
        //   If flip_grid_binding:
        //     If split_grid_y_dim:
        //       [i-remainder, BIDy{65535} | BIDx, TIDy | Unswitch, Unroll, TIDx]
        //     Else:
        //       [BIDy | BIDx, TIDy | Unswitch, Unroll, TIDx]
        //   Else:
        //     If split_grid_y_dim:
        //       [BIDx | i-remainder, BIDy{65535}, TIDy | Unswitch, Unroll, TIDx]
        //     Else:
        //       [BIDx | BIDy TIDy | Unswitch, Unroll, TIDx]
        // Else:
        //   If flip_grid_binding:
        //     If split_grid_y_dim:
        //       [i-remainder, BIDy{65535} | BIDx | Unswitch, Unroll, TIDx]
        //     Else:
        //       [BIDy | BIDx | Unswitch, Unroll, TIDx]
        //   Else:
        //     If split_grid_y_dim:
        //       [BIDx | i-remainder, BIDy{65535} | Unswitch, Unroll, TIDx]
        //     Else:
        //       [BIDx | BIDy | Unswitch, Unroll, TIDx]

        if (params.split_block) {
          reference_tv->split(1, NamedScalar::getParallelDim(ParallelType::TIDy));
          reference_tv->axis(2)->parallelize(ParallelType::TIDy);
        }
        auto bdimx_axis = (params.flip_grid_binding) ? 1 : 0;
        auto bdimy_axis = (params.flip_grid_binding) ? 0 : 1;
        reference_tv->axis(bdimx_axis)->parallelize(ParallelType::BIDx);
        if (params.split_grid_y_dim) {
          reference_tv->split(bdimy_axis, 65535);
          reference_tv->axis(bdimy_axis+1)->parallelize(ParallelType::BIDy);
        } else {
          reference_tv->axis(bdimy_axis)->parallelize(ParallelType::BIDy);
        }
```
I think the parallelization scheme could be condensed into something simple like this ^^^.

**Review by rdspring1 — APPROVED** [review]
LGTM. I left some minor comments.

**Review by naoyam — COMMENTED** [review]
LGTM

**Pull request merged**